### PR TITLE
Duplicate clones

### PR DESCRIPTION
**Pull request by BrianZanti**
* Prevents a student from filling out the clone form twice and creating duplicate clones
* Adds validations to Clone model to check for this condition
* Updates sad path so the student is notified when they create this condition
* Adds a clone show page where a student can see their clone and its status
* Links to the existing clone show page when a student tries to create a duplicate

Still to do:

* Check for this condition in the background worker (not sure if there's a scenario where this would be an issue)
* Only show link to repo on clone show page if the repo was created
* Probably makes for a better user experience if we alert the user that they already have this clone before they fill in the form. Does it make sense for the new clone link to redirect to the existing clone show page if the clone already exists?

### CODE REVIEW

File: app/views/clones/new.html.erb [hunk 1]
```diff
@@ -6,12 +6,12 @@
...
       <p class="mt2">
         <strong>Who will be working on this project?</strong>
-        <%= text_field_tag :students, "" , class: "block col-12 field"  %>
+        <%= text_field_tag :students, @clone.students, class: "block col-12 field"  %>
         <em>e.g. <strong>Leslie Knope</strong> if solo; <strong>Leslie Knope and Duke Silver.</strong> if a team project.</em>
       </p>
       <p class="mt2">
         <strong>We'll send you an email when everything is finished. Where would you like us to send to?</strong>
-        <%= text_field_tag :email, "" , class: "block col-12 field"  %>
+        <%= text_field_tag :email, nil, class: "block col-12 field"  %>
```
Comment (iandouglas) [inline app/views/clones/new.html.erb:14]: it'd be nice to populate this with `current_user.email` if we have it (it should be nil/empty otherwise), I think we only have one user in the production system who has a blank email address.

File: app/views/clones/show.html.erb [hunk 2]
```diff
@@ -0,0 +1,4 @@
+<h2>Clone of Project <%= @clone.project.name %></h2>
+<%= link_to 'View your repo on GitHub', @clone.url %>
+<p>Students working on this project: <%= @clone.students %></p>
+<p>Status: <%= @clone.message %></p>
```
Comment (iandouglas) [inline app/views/clones/show.html.erb:4]: It'd be nice to show the students what the different steps are and how far along it is in the process... not saying we have to do that here, but maybe as a future feature?

File: app/controllers/clones_controller.rb [hunk 3]
```diff
@@ -10,30 +10,42 @@ def new
...
   #   redirect_to admin_project_path(project), alert: "Cloning complete for #{clone.students}. Wait a minute and refresh the page."
   # end
+  private
 
   def clone_params
     params.require(:clone).permit(:url, :students)
   end
+
+  def clone_belongs_to_current_user?(clone)
+    clone.user == current_user
```
Comment (iandouglas) [inline app/controllers/clones_controller.rb:49]: Would it be worth adding `|| current_user.admin?` on here for anything?

File: spec/factories/projects.rb [hunk 4]
```diff
@@ -1,8 +1,8 @@
 FactoryBot.define do
   factory :project do
-    name { "Conveyor Belt" }
-    user { staff }
-    hash_id { "abc123" }
-    project_board_base_url { "https://github.com/turingschool-examples/watch-and-learn/projects/1" }
+    name { Faker::Company.catch_phrase }
+    user { create(:staff) }
```
Comment (iandouglas) [inline spec/factories/projects.rb:4]: does FactoryBot now require us to call `create` here instead of just the name of the factory? never seen this before.

File: spec/features/student/student_can_create_a_clone_spec.rb [hunk 5]
```diff
@@ -29,7 +23,7 @@
     expect(page).to have_content('Thanks for your submission!')
   end
 
-  scenario 'student cannot create a clone because ... reasons' do
+  scenario 'student cannot create a clone when that project has been deleted by an admin' do
```
Comment (iandouglas) [inline spec/features/student/student_can_create_a_clone_spec.rb:26]: but ... reasons ;)

**Review by iandouglas — COMMENTED** [review]
Good job on this, left some general comments that we can chat about on Slack

File: app/controllers/clones_controller.rb [hunk 6]
```diff
@@ -10,30 +10,42 @@ def new
...
+      @clone = @project.clones.new(students: params[:students], user: current_user, url: '')
+      if @clone.save
+        @clone.update(message: 'sending to sidekiq')
+        ProjectBoardClonerWorker.perform_later(@project, @clone, params[:email])
         redirect_to root_path, alert: "Thanks for your submission! We will send an email to #{params[:email]} when we finish getting everything setup. Follow the instructions in that message. Thanks!"
       else
-        redirect_to root_path, alert: "We're sorry but we were unable to clone the project board. If you continue to experience difficulty reach out to your instructor or point of contact."
+        existing_clone = Clone.find_by(user: current_user, project: @project)
+        click_here_link = helpers.link_to 'Click here', clone_path(existing_clone)
+        flash[:alert] = "We're sorry, it looks like you already have a clone created. #{click_here_link} to view your clone."
```
[the remark this answers is not in the document]
Comment (BrianZanti) [reply app/controllers/clones_controller.rb:23]: Yes! I was thinking to put that on the clone show page

File: app/models/clone.rb [hunk 7]
```diff
@@ -3,6 +3,8 @@ class Clone < ApplicationRecord
   belongs_to :user
 
   validates :students,  presence: true
+  validates :user, uniqueness: { scope: :project }
```
[the remark this answers is not in the document]
Comment (BrianZanti) [reply app/models/clone.rb:6]: I'll double check on this. The validation should only affect creates and updates on clones, so as long as those are still working we should be good.

File: app/views/clones/new.html.erb [hunk 8]
```diff
@@ -6,12 +6,12 @@
       <%= hidden_field_tag :project_hash_id, @project.hash_id  %>
       <p class="mt2">
         <strong>Who will be working on this project?</strong>
-        <%= text_field_tag :students, "" , class: "block col-12 field"  %>
+        <%= text_field_tag :students, @clone.students, class: "block col-12 field"  %>
```
[the remark this answers is not in the document]
Comment (BrianZanti) [reply app/views/clones/new.html.erb:9]: Yes, I will make that refactor 👍

File: app/views/clones/new.html.erb (hunk 1 above)
Comment (BrianZanti) [reply app/views/clones/new.html.erb:14]: I like that addition 👍

File: app/views/clones/show.html.erb (hunk 2 above)
Comment (BrianZanti) [reply app/views/clones/show.html.erb:4]: Yes, I think that would be great to add in to a future feature. Maybe a chance to add some cool JS? 😄

File: app/controllers/clones_controller.rb (hunk 3 above)
Comment (BrianZanti) [reply app/controllers/clones_controller.rb:49]: Yes, there's no reason an admin should be restricted from viewing this page.

File: spec/factories/projects.rb (hunk 4 above)
Comment (BrianZanti) [reply spec/factories/projects.rb:4]: I think you can use the name of the factory outside of the block, but not within the block. The error I get is `undefined method staff`.

File: spec/features/student/student_can_create_a_clone_spec.rb (hunk 5 above)
Comment (BrianZanti) [reply spec/features/student/student_can_create_a_clone_spec.rb:26]: I changed it because... reasons.

File: app/controllers/clones_controller.rb [hunk 9]
```diff
@@ -10,30 +10,42 @@ def new
...
+        render :new
       end
     else
       redirect_to root_path, alert: "We're sorry but we were unable to clone the project board."
     end
   end
 
+  def show
+    @clone = Clone.find(params[:id])
```
[the remark this answers is not in the document]
Comment (BrianZanti) [reply app/controllers/clones_controller.rb:32]: Yes, that's a good edge case test
